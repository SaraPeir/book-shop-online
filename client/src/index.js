import React from 'react';
import ReactDOM from 'react-dom';
import './index.scss';
import App from './App';
import 'bootstrap/dist/css/bootstrap.min.css';
import { ApolloClient, HttpLink, InMemoryCache, ApolloProvider  } from '@apollo/client';

const client = new ApolloClient({
  cache: new InMemoryCache(),
  link: new HttpLink({
    uri: 'http://localhost:4000/graphql'
  })
});

console.log('client', client);

ReactDOM.render(
  <ApolloProvider client={client}>
    <App />
  </ApolloProvider>,
  document.getElementById('root')
);
