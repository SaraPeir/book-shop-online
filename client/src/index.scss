@import "variables.scss";

.main-style {
    color: $primary-color;
}