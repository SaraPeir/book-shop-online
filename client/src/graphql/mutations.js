import { gql } from '@apollo/client';

export const UPDATE_BOOKS = gql`
    mutation updateBook($id: ID, $isFavourite: Boolean) {
        updateBook(input: {
        id: $id,
        isFavourite: $isFavourite
        }) {
            title
            description
            id
            author
            price
            score
            count
            pages
            image
            isFavourite
            type
        }
    }
`