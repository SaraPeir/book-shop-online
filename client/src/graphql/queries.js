import { gql } from '@apollo/client';

export const GET_BOOKS  = gql`
query getBooks {
    getBooks {
      title
      description
      id
      author
      price
      score
      count
      pages
      image
      isFavourite
      type
    }
  }
` 