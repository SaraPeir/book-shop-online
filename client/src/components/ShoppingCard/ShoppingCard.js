import React from 'react';
import { useMutation } from '@apollo/client';
import {
    Card, CardImg, CardText, CardBody,
    CardTitle, CardSubtitle, Button
  } from 'reactstrap';

import {Heart} from '../../images/icons/index.js';
import {UPDATE_BOOKS} from '../../graphql/mutations';
import {GET_BOOKS} from '../../graphql/queries.js';

// onCompleted usalo per creare un toaster quando l'operazione di mtation è completa:
// onCompleted

// añadir lazyloading

const ShoppingCard = (props) => {
  const [updateBooks] = useMutation(
      UPDATE_BOOKS,
      {
          refetchQueries: () => [
            {
                query: GET_BOOKS
              }
          ]
      }, // para actualizar la cache instantaneamente
    );

      return (
        <Card className={props.cardStyle}>
            {props.isForFavourites && 
            <div className="close-button-icon-container"
            onClick={() => {
                props.updateBooks(); // a lo mejor para meter un toaster
                updateBooks({
                    variables: {
                        id: props.id, 
                        isFavourite: !props.isFavourite
                    }
                })
            }} >
                X
            </div>
            }
            <div className={"img-container"}>
            {!props.isForFavourites && 
                <div 
                    className={`${props.isFavourite ? "filled-icon icon-container" : "icon-container"}` }
                    id={"heart-button"} 
                    onClick={() => {
                        props.updateBooks(); // a lo mejor para meter un toaster
                        updateBooks({
                            variables: {
                            id: props.id, 
                            isFavourite: !props.isFavourite
                            } 
                        })
                    }}>
                    {Heart}
                </div>
            }
                <CardImg src={props.image} alt="Card image cap" />
            </div>
            <CardBody>
                <CardTitle>{props.title}</CardTitle>
                <CardSubtitle>{props.author}</CardSubtitle>
                
                <div className={"price-container"}>
                <CardText>{props.price}</CardText>
                </div>
                <div className={"btn-container"}>
                    <Button>Dettagli</Button>
                </div>
            </CardBody>
        </Card> 
      );
  }

  export default ShoppingCard;