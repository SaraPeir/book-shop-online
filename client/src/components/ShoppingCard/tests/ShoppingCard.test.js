import React from 'react';
import ShoppingCard from '../ShoppingCard';
import { MockedProvider } from '@apollo/client/testing';
import {UPDATE_BOOKS} from '../../../graphql/mutations';
import {GET_BOOKS} from '../../../graphql/queries.js';

jest.mock('pure-react-carousel')

import { render, 
   fireEvent, 
   waitFor,
    screen } from '@testing-library/react'

    // npm test -- --coverage --watchAll=false para el report
    const mockedFn = jest.fn();

    const mocks = [
        {
          request: {
            query: UPDATE_BOOKS,
            variables: {
              id: "5eb83d4b7e177a8a54277468",
              isFavourite: true
            },
          },
          result: {
            data: {
                updateBook: {
                id: "5eb83d4b7e177a8a54277468",
                isFavourite: true
              },
            },
          }
       
        },
        {
            request: {
              query: GET_BOOKS, // necessary to mock refetching after mutation
              variables: {},
            },
            result: {
              data: {
                  getBook: [
                    {
                        author: "AUTHOR_1",
                        count: 3,
                        description: "BOOK1_DESCIPTION.",
                        id: "5eb83d4b7e177a8a54277468",
                        image: "IMG_1",
                        isFavourite: false,
                        pages: 312,
                        price: "€ 18.00",
                        score: 4,
                        title: "TITLE_1",
                        type: "TYPE_1"
                    }
                  ],
              },
            }
          },
      ]

const data = {
    cachedBook: [
        {
            author: "AUTHOR_1",
            count: 3,
            description: "BOOK1_DESCIPTION.",
            id: "5eb83d4b7e177a8a54277468",
            image: "IMG_1",
            isFavourite: true,
            pages: 312,
            price: "€ 18.00",
            score: 4,
            title: "TITLE_1",
            type: "TYPE_1"
        }
    ]
}

const props = {
    author: "AUTHOR_1",
    count: 3,
    description: "BOOK1_DESCIPTION.",
    id: "5eb83d4b7e177a8a54277468",
    image: "IMG_1",
    isFavourite: false,
    pages: 312,
    price: "€ 18.00",
    score: 4,
    title: "TITLE_1",
    type: "TYPE_1"
}
it('Should render correctly when isForFavourites', () => {
    const {container, getByTestId} = render(
        <MockedProvider mocks={mocks}>
            <ShoppingCard {...props}  />
        </MockedProvider>,
        
      );

      const closeButton = getByTestId('heart-button');
      expect(closeButton).toBeInTheDocument()
  
      expect(container).toMatchSnapshot()
})

it('Should render correctly when isForFavourites is true', () => {
    const {container, getByTestId} = render(
        <MockedProvider mocks={mocks}>
            <ShoppingCard {...props} isForFavourites={true}  />
        </MockedProvider>,
        
      );

      const closeButton = getByTestId('close-button');
      expect(closeButton).toBeInTheDocument()
  
      expect(container).toMatchSnapshot()
})

it('Should execute a function when you click at heart button', () => {
    let button;
    const {container, getByTestId, rerender} = render(
        <MockedProvider mocks={mocks}>
            <ShoppingCard {...props} updateBooks={mockedFn} />
        </MockedProvider>
      );
      button = getByTestId('heart-button');
      expect(button).toBeInTheDocument()
  
      //   // solucion desde aquí: https://jkettmann.com/testing-apollo-how-to-test-if-a-mutation-was-called-with-mockedprovider/
    //   const addBookMutationMock = mocks[0].result;
    //   addBookMutationMock()
    //   expect(addBookMutationMock).toHaveBeenCalled()
        fireEvent.click(button);

    setTimeout(() => expect(mockedFn).toHaveBeenCalled(), 100);
})

// { wrapper: CarouselSlide}

// it('Should render correctly', () => {
//     const {container} = render(
//             <CarouselCard {...props}  />
//       );
  
//       expect(container).toMatchSnapshot()
// })