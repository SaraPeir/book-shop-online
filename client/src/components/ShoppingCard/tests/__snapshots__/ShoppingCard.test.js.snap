// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render correctly when isForFavourites 1`] = `
<div>
  <div
    class="card"
  >
    <div
      class="img-container"
    >
      <div
        class="icon-container"
        data-testid="heart-button"
        id="heart-button"
      >
        <svg
          height="24"
          stroke="red"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 4.248c-3.148-5.402-12-3.825-12 2.944 0 4.661 5.571 9.427 12 15.808 6.43-6.381 12-11.147 12-15.808 0-6.792-8.875-8.306-12-2.944z"
          />
        </svg>
      </div>
      <img
        alt="Card image cap"
        class="card-img"
        src="IMG_1"
      />
    </div>
    <div
      class="card-body"
    >
      <div
        class="card-title"
      >
        TITLE_1
      </div>
      <div
        class="card-subtitle"
      >
        AUTHOR_1
      </div>
      <div
        class="price-container"
      >
        <p
          class="card-text"
        >
          € 18.00
        </p>
      </div>
      <div
        class="btn-container"
      >
        <button
          class="btn btn-secondary"
        >
          Dettagli
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Should render correctly when isForFavourites is true 1`] = `
<div>
  <div
    class="card"
  >
    <div
      class="close-button-icon-container"
      data-testid="close-button"
    >
      X
    </div>
    <div
      class="img-container"
    >
      <img
        alt="Card image cap"
        class="card-img"
        src="IMG_1"
      />
    </div>
    <div
      class="card-body"
    >
      <div
        class="card-title"
      >
        TITLE_1
      </div>
      <div
        class="card-subtitle"
      >
        AUTHOR_1
      </div>
      <div
        class="price-container"
      >
        <p
          class="card-text"
        >
          € 18.00
        </p>
      </div>
      <div
        class="btn-container"
      >
        <button
          class="btn btn-secondary"
        >
          Dettagli
        </button>
      </div>
    </div>
  </div>
</div>
`;
